smithy-model/src/main/java/software/amazon/smithy/model/transform/DeconflictErrorsWithSharedStatusCode.java

```diff
@@ -0,0 +1,158 @@
+            if (replaceOperation) {
+                replacementOperation.build();
+                shapesToReplace.add(replacementOperation.build());
+            }
+        }
+
+        return transformer.replaceShapes(model, shapesToReplace);
+    }
+
+    // Return synthetic error, along with any updated shapes.
+    private Pair<Shape, List<Shape>> synthesizeErrorUnion(String operationName, String statusCode,
+                                                          List<StructureShape> errors) {
+        List<Shape> replacementShapes = new ArrayList<>();
+        StructureShape.Builder errorResponse = StructureShape.builder();
+        ShapeId errorResponseId = ShapeId.fromParts(forService.getId().getNamespace(),
+                operationName + statusCode + "Error");
+        errorResponse.id(errorResponseId);
+        errorResponse.addTraits(getErrorTraitsFromStatusCode(statusCode));
+        Map<String, HttpHeaderTrait> headerTraitMap = new HashMap<>();
+        UnionShape.Builder errorUnion = UnionShape.builder().id(
+                ShapeId.fromParts(errorResponseId.getNamespace(), errorResponseId.getName() + "Content"));
+        for (StructureShape error : errors) {
+            StructureShape newError = createNewError(error, headerTraitMap);
+            replacementShapes.add(newError);
+            MemberShape newErrorMember = MemberShape.builder()
+                    .id(errorUnion.getId().withMember(newError.getId().getName()))
+                    .target(newError.getId())
+                    .build();
+            replacementShapes.add(newErrorMember);
+            errorUnion.addMember(newErrorMember);
+        }
+        UnionShape union = errorUnion.build();
+        replacementShapes.add(union);
+        errorResponse.addMember(MemberShape.builder()
+                .id(errorResponseId.withMember("errorUnion"))
+                .target(union.getId())
+                .build());
+        // Add members with hoisted HttpHeader traits.
+        for (Map.Entry<String, HttpHeaderTrait> entry : headerTraitMap.entrySet()) {
+            errorResponse.addMember(MemberShape.builder().id(errorResponseId.withMember(entry.getKey()))
+                    .addTrait(entry.getValue()).target("smithy.api#String").build());
+        }
+        StructureShape built = errorResponse.build();
+        return Pair.of(built, replacementShapes);
+    }
+
+    private StructureShape createNewError(StructureShape oldError, Map<String, HttpHeaderTrait> headerMap) {
+        StructureShape.Builder newErrorBuilder = oldError.toBuilder().clearMembers();
+        for (MemberShape member : oldError.getAllMembers().values()) {
+            String name = member.getMemberName();
+            // Collect HttpHeaderTraits to hoist.
+            if (member.hasTrait(HttpHeaderTrait.ID)) {
+                HttpHeaderTrait trait = member.expectTrait(HttpHeaderTrait.class);
+                if (headerMap.containsKey(name)) {
```

Comment:
thought: You could put the value and check the return if it is null to reduce some `if` nesting.
```
HttpHeaderTrait newTrait = member.expectTrait(HttpHeaderTrait.class);
HttpHeaderTrait previousTrait = headerMap.put(name, newTrait);
if (previousTrait != null && previousTrait.equals(newTrait)) {
    throw MyError
}
```